packages/hebao_v2/contracts/base/libwallet/MetaTxLib.sol

```diff
@@ -60,17 +61,22 @@ library MetaTxLib
         view
         returns (bytes32)
     {
+        // If this is a dataless meta-tx the user only signs the function selector,
+        // not the full function calldata.
+        bytes memory data = metaTx.nonce == 0 ? metaTx.data.slice(0, 4) : metaTx.data;
```

Comment:
I still think there is a security issue here. Assuming two different recovery requests have been signed by the majority of all guardians, now the owner doesn't really know which recovery will be executed and it's totally up to the operator to decide what payload should be injected into the meta transaction.

Reply:
Can we do this instead:
1. inside xxxWA methods, we return the hash signed by guardians
2. in executeMetaTx, somewhere around `(success, ) = metaTx.to.call{gas: metaTx.gasLimit}(metaTx.data);` we try to extract this **txAwareHash**, if we find one, then we pass it into validateMetaTx and request the owner also sigh the hash together with other fee related parameters. If this hash is not found, we use 0x0.

Reply:
I don't believe that's a problem. If the owner wants to pay a certain fee for an operation, than it doesn't really matter if this is done to A, B or both. It's only really about what fee is acceptable by the owner for a certain operation for some period of time, and that's it. And for all these *WA operations (except `recover`, but the guardians can call this directly to get all the funds in the wallet directly) the requirement is `MAJORITY_OWNER_REQUIRED` which means the owner actually does have to approve the use of the meta-tx indirectly as well. So without the owner directly involved it's impossible to actually create new valid transactions using the same meta-tx.

Getting the hash as the return value is a good idea, if you thing this is a problem that's a good cheap solution as well.